content/cucumber/reporting.md

```diff
@@ -235,6 +236,137 @@ Alternatively, you can specify a Java system property (in Maven, Gradle or other
 ```
 {{% /block %}}
 
+# Custom formatters
+The Cucumber formatter API is readily extensible. A formatter can be any class implementing the event-based formatter
+API. {{% text "ruby" %}}The formatter class should live in the `features/support` directory. {{% /text %}}To use your custom formatter, run Cucumber using the `--format` flag:
+```
+cucumber --format CustomFormatter
+```
+
+## Formatter API
+Cucumber uses an event-based API for its formatters. These formatters respond to several defined events, with
+event handlers defined in the formatter's constructor. 
+{{% block "ruby" %}}
+A sample formatter could look like this:
+```ruby
+class CustomFormatter
+  attr_reader :io
+
+  def initialize(config)
+    @io = config.out_stream
+
+    # Using a block
+    config.on_event :test_step_started do |event|
+      io.puts 'Started test step'
+    end
+
+    # Passing in a method
+    config.on_event :test_step_finished &method(:print_finished)
+  end
+
+  def print_finished(event)
+    io.puts 'Finished test step'
+  end
+end
+```
+{{% /block %}}
+
+### Configuration object
+The formatter initializer is passed a {{% text "java,dotnet" %}}Cucumber configuration{{% /text %}}{{% text "ruby" %}}`Cucumber::Configuration`{{% /text %}} object. This is the 
+[configuration for the test run](https://docs.cucumber.io/cucumber/configuration/),
+including default configurations and [options](https://docs.cucumber.io/cucumber/api/) passed in at the command line.
+It can be useful to access these options, so that your formatter can modify its behavior in response to user directives.
+
+### Event objects
+Every time an event is fired, an event object is passed to the corresponding handler. Each event object belongs to
+a class corresponding to its name. {{% text "ruby" %}}For instance, the `gherkin_source_read` event creates a
```

Comment:
I've added additional "block" and "text" tags to some (but not yet all!) things that are obviously Ruby specific. I'd like to have some people who have more knowledge of the Java/Javascript/Dotnet implementations of formatters check that what I think is generic for all languages is *in fact correct* for all languages. (@mpkorstanje would you be willing to help for the Java part, please?)